### ISSUE TITLE
Normalizing Flow When transform style is MAF(Masked Autoregressive Flow)

### ISSUE THREAD

**Issue by GongsonChan** [GongsonChan/code#16]
## Library
```
import pandas as pd
import torch
from sklearn.preprocessing import StandardScaler
from nflows.flows.base import Flow
from nflows.transforms.normalization import BatchNorm
from nflows.transforms.autoregressive import MaskedAffineAutoregressiveTransform
from nflows.transforms.permutations import ReversePermutation
from nflows.transforms.base import CompositeTransform
from nflows.distributions.normal import StandardNormal
```

## Data
```
df = pd.read_csv('C:/Users/hhccw/OneDrive/바탕 화면/통공모/tgm2025_data.csv')  # 또는 df = your_dataframe
data = df[['lon', 'lat', 'time', 'intensity']].values
```

## Scailing
```
scaler = StandardScaler()
data_scaled = scaler.fit_transform(data)
```

## Transforming to Torch Tensor
`x = torch.tensor(data_scaled, dtype=torch.float32)`

## Model Organization
```
num_features = x.shape[1]
hidden_features = 6
num_layers = 4
```

##  Flow Model Organization
```
transforms = []
for _ in range(num_layers):
    transforms.append(MaskedAffineAutoregressiveTransform(
        features=num_features, hidden_features=hidden_features)) # 비선형, 확률 밀도 변경 변환
    transforms.append(BatchNorm(num_features)) # 각 차원의 평균과 분산 정규화 and log-determinant도 계산 가능
    transforms.append(ReversePermutation(features=num_features)) # 변수 순서 뒤집기

transform = CompositeTransform(transforms) # transforms를 하나의 복합 함수로 coposite
base_dist = StandardNormal(shape=[num_features]) # 각 변수를 정규분포에서 sampling(z ~ N(0, I))
flow = Flow(transform, base_dist) # 전체 Flow 구성
```

## 파라미터 수 확인
```
total_params = sum(p.numel() for p in flow.parameters())
print(f"모형의 전체 파라미터 수: {total_params}")
```
>> 모형의 전체 파라미터 수: 993 - 조건 충족

## Model Training
```
optimizer = torch.optim.Adam(flow.parameters(), lr=1e-3) # Adam 최적화 알고리즘 그리고 학습률
num_epochs = 1000 # 에포크: 반복횟수

for epoch in range(num_epochs):
    optimizer.zero_grad() # gradient 초기화
    loss = -flow.log_prob(x).mean()  # Negative Log-Likelihood - 딥러닝은 loss를 최소화 -> 부호 반대로
    loss.backward() # 역전파 - 모든 파라미터의 gradient를 자동으로 계산
    optimizer.step() # 모델 파라미터 업데이트
    if epoch % 100 == 0:
        print(f"Epoch {epoch}, Loss: {loss.item():.4f}")
```
## 로그 가능도 최대화 방법

설정한 Normalizing Flow 모형은 닫힌 형태의 로그 가능도 함수 최적해를 구할 수 없기 때문에, 수치적 최적화 방법인 **경사상승법 (Gradient Ascent)**을 사용합니다.

PyTorch에서는 이를 역전파(Backpropagation)를 통해 구현하며, 손실 함수로 Negative Log-Likelihood (NLL)를 사용하여 다음과 같은 방식으로 최적화합니다:

<img width="292" height="75" alt="Image" src="https://github.com/user-attachments/assets/db944a10-e154-4c54-a393-90c8ff0f3541" />

이 손실 함수를 최소화하는 것이 곧 로그 가능도를 최대화하는 것과 동치입니다.
Optimizer는 Adam을 사용하며, 학습률은 0.001로 설정하였습니다.

## 로그 가능도 함수의 정의

설정한 Normalizing Flow 모형에서 로그 가능도 함수는 다음과 같이 정의됩니다:

<img width="583" height="312" alt="Image" src="https://github.com/user-attachments/assets/6c567bf0-302b-48ce-bfeb-85d9fb3a4a20" />

## Log Likelihood
```
log_likelihood = flow.log_prob(x).sum().item()
print(f"총 Log-Likelihood: {log_likelihood:.4f}")
```
>> 총 Log-Likelihood: -94637.0938
